runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java

```diff
@@ -672,11 +706,84 @@ public void freeTableData(String namespace, String tableName) {
+     *
+     * @param namespace Namespace of the table.
+     * @param tableName Name of the table.
+     */
+    public void dropTable(String namespace, String tableName) {
+        log.info("Invoking dropTable {}${}", namespace, tableName);
+        clearAndUnregisterTable(namespace, tableName);
+
+        String fullyQualifiedTableName = getFullyQualifiedTableName(namespace, tableName);
+        tableMap.remove(fullyQualifiedTableName);
+        log.info("Done dropping table {}${}", namespace, tableName);
+    }
+
+    private void clearAndUnregisterTable(String namespace, String tableName) {
+        TableName tableNameKey = TableName.newBuilder()
+                .setNamespace(namespace)
+                .setTableName(tableName)
+                .build();
+
+        // Mark this table as dropped in RegistryTable
+        int numRetries = 9; // Since this is an internal transaction, retry a few times before giving up.
+        while (numRetries-- > 0) {
+            if (TransactionalContext.isInTransaction()) {
+                throw new IllegalThreadStateException("openTable: Called on an existing transaction");
+            }
+            try {
+                this.runtime.getObjectsView().TXBuild()
+                        .type(TransactionType.WRITE_AFTER_WRITE)
+                        .build()
+                        .begin();
+
+                // Clear table data
+                UUID streamId = CorfuRuntime.getStreamID(getFullyQualifiedTableName(namespace, tableName));
+                SMREntry entry = new SMREntry("clear", new Array[0], Serializers.PRIMITIVE);
+                TransactionalContext.getCurrentContext().logUpdate(streamId, entry);
+
+                UUID cpStreamId = CorfuRuntime.getCheckpointStreamIdFromId(streamId);
+                TransactionalContext.getCurrentContext().addToDeleteSet(streamId);
+                TransactionalContext.getCurrentContext().addToDeleteSet(cpStreamId);
+
+                CorfuRecord<TableDescriptors, TableMetadata> oldRecord =
+                        this.registryTable.get(tableNameKey);
+                if (oldRecord == null || oldRecord.getPayload() == null) {
+                    log.warn("The table to unregister {} does not exist in RegistryTable.", tableNameKey);
+                } else {
+                    CorfuRecord<TableDescriptors, TableMetadata> newRecord = new CorfuRecord<>(
+                            oldRecord.getPayload(),
+                            TableMetadata.newBuilder().mergeFrom(oldRecord.getMetadata()).setIsDropped(true).build());
+                    this.registryTable.insert(tableNameKey, newRecord);
+                }
+
+                this.runtime.getObjectsView().TXEnd();
+
+                log.info("Successfully cleared and unregister table {}${}", namespace, tableName);
+                break;
+            } catch (TransactionAbortedException txAbort) {
+                if (numRetries <= 0) {
```

Comment:
This code assumes that side-effect of `addToDeleteSet` and the transaction commit phase to be atomic, but it isn't.

For example, the sequencer can delete the streams, but the write can fail leaving the system in an inconsistent state.